proposals/p1927.md

```diff
@@ -0,0 +1,148 @@
+# Slice generator
+
+<!--
+Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+Exceptions. See /LICENSE for license information.
+SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+-->
+
+[Pull request](https://github.com/carbon-language/carbon-lang/pull/1927)
+
+<!-- toc -->
+
+## Table of contents
+
+-   [Problem](#problem)
+-   [Background](#background)
+-   [Proposal](#proposal)
+-   [Details](#details)
+    -   [Library API](#library-api)
+-   [Rationale](#rationale)
+-   [Alternatives considered](#alternatives-considered)
+
+<!-- tocstop -->
+
+## Problem
+
+Slices of arrays is a popular way to extract subarray of array. They also may be
+used for generating arrays.
+
+## Background
+
+There is no requirements for background.
+
+## Proposal
+
+Introduce slice generators.
+
+## Details
+
+Slice syntax is `a:b:s`. Here, `a` is the beginning of slice, `b` is the end of
+slice and `s` is the step in slice. In array indexing, `:s` can be omitted, then
+it means that step is 1. If `a` is omitted, slice starts from the beginning of
+array. The same situation is for `b`, but slice ends at the end of arrays. `:`
+is the same as `::`. If `s` is negative and `a` or `b` or both are skipped,
+instead of `a`, upper bound is and, instead of `b`, lower bound is. Slice has
+closed range.
+
+Compilation or runtime errors are:
+
+-   if `s` is zero;
+-   if `a` is greater than `b` when `s` is positive;
+-   if `a` is less than `b` when `s` is negative.
+
+`:` has the lowest priority for making possible `a[1:N-1]` without any brackets.
+
+Slice generator may return `i32`, `i64`, `i128` and unsigned versions of them.
+
+Slice as generator of arrays:
+
+```carbon
+  var array1: auto = (1:5:2); // array1 = (1,3,5)
+  var array2: auto = (1:5); // array1 = (1,2,3,4,5)
+```
+
+Round brackets, `a`, and `b` must be presented, `s` may be omitted,
+
+In operator `[]` of array, slices can be used in the following ways:
+
+```
+  var array: auto = (1:5:2); // array = (1,3,5)
```

Comment:
This seems likely to be an efficiency surprise. As with python's `range`, I think we'll want a slice generator to lazily  produce the elements of the slice. Having a conversion from slice generator to array seems like it might be reasonable, but only if the developer somehow asks for it (eg, by writing an array type).